api/v1alpha7/nnfcontainerprofile_types.go

```diff
@@ -69,33 +69,114 @@ type NnfContainerProfileData struct {
 	GroupID *uint32 `json:"groupID,omitempty"`
 
-	// Number of ports to open for communication with the user container. These ports are opened on
-	// the targeted NNF nodes and can be accessed outside of the k8s cluster (e.g. compute nodes).
-	// The requested ports are made available as environment variables inside the container and in
-	// the DWS workflow (NNF_CONTAINER_PORTS).
+	// Number of ports to open for each container specified in the PodSpec. For MPI Jobs, this is
+	// only for the Launcher container(s) listed in the MPIReplicaSet's PodSpec. These ports are
+	// opened on the targeted NNF nodes and can be accessed outside the k8s cluster (e.g. compute
+	// nodes). The requested ports are made available as environment variables inside the container
+	// and in the DWS workflow (NNF_CONTAINER_PORTS).
 	NumPorts int32 `json:"numPorts,omitempty"`
 
-	// Spec to define the containers created from this profile. This is used for non-MPI containers.
-	// Refer to the K8s documentation for `PodSpec` for more definition:
-	// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodSpec
-	// Either this or MPISpec must be provided, but not both.
-	Spec *corev1.PodSpec `json:"spec,omitempty"`
-
-	// MPIJobSpec to define the MPI containers created from this profile. This functionality is
-	// provided via mpi-operator, a 3rd party tool to assist in running MPI applications across
-	// worker containers.
-	// Either this or Spec must be provided, but not both.
-	//
-	// All the fields defined drive mpi-operator behavior. See the type definition of MPISpec for
-	// more detail:
-	// https://github.com/kubeflow/mpi-operator/blob/v0.4.0/pkg/apis/kubeflow/v2beta1/types.go#L137
-	//
-	// Note: most of these fields are fully customizable with a few exceptions. These fields are
-	// overridden by NNF software to ensure proper behavior to interface with the DWS workflow
-	// - Replicas
-	// - RunPolicy.BackoffLimit (this is set above by `RetryLimit`)
-	// - Worker/Launcher.RestartPolicy
-	MPISpec *mpiv2beta1.MPIJobSpec `json:"mpiSpec,omitempty"`
+	// NnfSpec to define the containers created from this profile. This is used for non-MPI containers.
+	// Either this or NnfMPISpec must be provided, but not both.
+	// +kubebuilder:validation:Rule="(self.spec != null) != (self.mpiSpec != null)",Message="Exactly one of 'spec' or 'mpiSpec' must be set."
+	NnfSpec *NnfPodSpec `json:"spec,omitempty"`
+
+	// MPIJobSpec to define the MPI containers created from this profile.
+	// Either this or NnfSpec must be provided, but not both.
+	// +kubebuilder:validation:Rule="(self.spec != null) != (self.mpiSpec != null)",Message="Exactly one of 'spec' or 'mpiSpec' must be set."
+	NnfMPISpec *NnfMPISpec `json:"mpiSpec,omitempty"`
+}
+
+// NnfPodSpec represents the specification of a pod that can be used in a container profile. This is
+// a slimmed down version of a corev1.PodSpec to reduce the size of the CRD.
+type NnfPodSpec struct {
```

Comment:
You're not representing the 'mpiReplicaSpecs' layer.

```
$ kubectl apply -f environments/kind/nnf-dm/reference/copy-offload-template-nnfcontainerprofile.yaml
Error from server (BadRequest): error when creating "environments/kind/nnf-dm/reference/copy-offload-template-nnfcontainerprofile.yaml": NnfContainerProfile in version "v1alpha7" cannot be handled as a NnfContainerProfile: strict decoding error: unknown field "data.mpiSpec.mpiReplicaSpecs", unknown field "data.mpiSpec.runPolicy"
```

We could make upgrades easier by representing that layer, so the customer doesn't have to edit these yaml files, and fix the indentation, and probably get it wrong the first time so they have to do it again.

Reply:
Wow, never mind. The existing yamls need big changes anyway.

This wasn't caught by unpack-manifests.  I'll have to look at that.